keps/sig-node/2535-ensure-secret-pulled-images/README.md

```diff
@@ -129,27 +136,25 @@ use un-encrypted...
 
 This feature will not change the behavior of pod with `ImagePullPolicy` `Always`.
 
+Enforcing periodical repull might be important in order to, for example, be able
+to recheck image license entitlement, but is a non-goal for this enhancement.
+
 ## Proposal
 
-For alpha the kubelet will keep a list, across reboots of host and restart of
-kubelet, of container images that required authentication and a list of the
-authentications that successfully pulled the image.
-For beta an API will be considered to manage the ensure metadata.
+A new API will be created to represent metadata about image pulls being authenticated. This API
+will be used by the kubelet to persist this metadata in file(s) on the kubelet's node.
+The API is a representation of the kubelet's in-memory state, where the kubelet tracks
+information about image pulls being authenticated.
 
-The kubelet will ensure any image in the list is always pulled if an authentication
-used is not present, thus enforcing authentication / re-authentication.
+The kubelet will ensure any image it keeps a track of is always pulled if authentication information
+from an image pull is not yet present, thus enforcing authentication / re-authentication.
 
-There will be two different kubelet configuration options added, as well as a feature
-gate to gate their use:
-- `pullImageSecretRecheck`
-    - A boolean that toggles this behavior. If `false`, the kubelet will fallback to the
+A new kubelet configuration option will be added, as well as a feature
+gate to gate its use:
+- `noPullImageSecretRecheck`
```

Comment:
let's noodle a little bit on whether we can come up with a config knob / knobs that makes the spectrum of enforcement coherent:

How would each of these use cases be expressed in the kubelet config?
```
1. Do not allow access to _any_ pre-loaded images without verification
2. Only allow access to a _specific_ list of pre-loaded images without verification
3. Allow access to _all_ pre-loaded and pulled-without-pull-secrets images without verification (default behavior when this feature is enabled)
4. Allow access to all images _without verification_ (default behavior when this feature is not enabled, not recommended)
```

(again, I'm not 100% sure we want to make number 4 possible)

strawman possibilities (I didn't spend a ton of time on this, just wanted to start you thinking):

For 1:
```yaml
preexistingImagePolicy: VerifyAccess
```

For 2:
```yaml
preexistingImagePolicy: VerifyAccessUnlessAllowlisted
preexistingImageAllowlist:
- myregistry.example.com/myimage:mytag
- myregistry.example.com/myotherimage:*
- mypublicregistry.example.com/*
```

For 3 (default policy):
```yaml
preexistingImagePolicy: AllowAccessWithoutVerification
```

Reply:
Given the scenarios we've got, that's actually a good option list. I think I'd probably only name it differently, maybe structure it a bit more for simpler config validation:
```yaml
disableImageSecretPullRecheck: true | false # in case we actually want to be able to disable this
preexistingImagesPullPolicy:
    type: AlwaysVerify | IgnoreListed | NeverVerify
    ignoreList: # specific to image tag / just an image + all tags / whole registry
    - myregistry.example.com/myimage:mytag
    - myregistry.example.com/myotherimage:*
    - mypublicregistry.example.com/*
```
```go
type KubeletConfiguration struct {
...
   DisableImageSecretPullRecheck bool // defaults to false
   PreexistingImagesPullPolicy *PreexistingImagesSecretPullPolicy
...
}

type PreexistingImageAccessPolicy string

const (
    NeverVerifyPreexistingImages PreexistingImageAccessPolicy = "NeverVerify" // the default
    IgnoreListedPreexistingImages = "IgnoreListed"
    AlwaysVerifyPreexistingImages = "AlwaysVerify"
)

type PreexistingImagesSecretPullPolicy struct {
    Type PreexistingImageAccessPolicy
    IgnoreList []string
}
```

Reply:
This was further noodled about during a meeting with sig-node yesterday, I'll work in the sketch we came up with.